pageserver/src/walredo.rs

```diff
@@ -90,16 +98,22 @@ pub trait WalRedoManager: Send + Sync {
 
 ///
 /// This is the real implementation that uses a Postgres process to
-/// perform WAL replay. Only one thread can use the process at a time,
-/// that is controlled by the Mutex. In the future, we might want to
-/// launch a pool of processes to allow concurrent replay of multiple
-/// records.
+/// perform WAL replay. It multiplexes requests from multiple threads
+/// using `sender` channel and send them to the postgres wal-redo process
+/// pipe by separate thread. Responses are returned through set of `receivers`
+/// channels, used in round robin manner.  Receiver thread is protected by mutex
+/// to prevent it's usage by more than one thread
+/// In the future, we might want to launch a pool of processes to allow concurrent
+/// replay of multiple records.
 ///
 pub struct PostgresRedoManager {
-    tenant_id: TenantId,
-    conf: &'static PageServerConf,
-
-    process: Mutex<Option<PostgresRedoProcess>>,
+    // mutiplexor pipe: use sync_channel to allow sharing sender by multiple threads
+    // and limit size of buffer
+    sender: SyncSender<(ChannelId, Vec<u8>)>,
+    // set of receiver channels
+    receivers: Vec<Mutex<Receiver<Bytes>>>,
+    // atomicly incremented counter for choosing receiver
+    round_robin: AtomicUsize,
```

Comment:
The usual way to have a request/response queue like this is to create a "oneshot" channel for each request, and send the sending half of it along with the request. So something like:

```
    /// To request WAL redo, create a oneshot channel for the response, and send the
    /// serialized request and the send-half of the oneshot channel here. Then wait for
    /// the response on the receive-half of the oneshot channel
    request_queue: tokio::mpsc::Sender<(tokio::sync::oneshot::Sender, Vec<u8>),
```
You can use those tokio channels from blocking code too, the Senders and Receivers also have blocking send and receive functions.

Reply:
With `blocking_recv` I get the following error:
```
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.
```
I wonder why it is necessary to use `tokio::sync::oneshot::Sender` ?
I tried standard `std::sync::mpsc::Sender` and speed is the same.
I have tested speed on such simple test:
```
    let (tx, rx): (
        SyncSender<(Sender<u32>,u32)>,
        Receiver<(Sender<u32>,u32)>,
    ) = mpsc::sync_channel(1024*1024);
    std::thread::spawn(move || {
		for i in 0..N_ITER {
			let (sender,data) = rx.recv().unwrap();
			assert!(data == i);
			sender.send(data).unwrap();
		}
	});

	for i in 0..N_ITER {
		let (tx1, rx1) = mpsc::channel();
		tx.send((tx1,i)).unwrap();
		let resp = rx1.recv().unwrap();
		assert!(resp == i);
	}
```

Reply:
I also tried  https://docs.rs/crossbeam-channel/0.5.6/crossbeam_channel/index.html
because I saw recommendation to use it as much faster alternative to std::sync::mpsc.
But at my test I didn't see significant difference.

Reply:
> I wonder why it is necessary to use tokio::sync::oneshot::Sender ?
> I tried standard std::sync::mpsc::Sender and speed is the same.

This is just a design pattern. With oneshot it's clear at the type level that you are just going to get back only a single response. The `std`'s rather complex mpsc optimizes for this "oneshot" usage first, then switches to more different kind of implementation when `Sender`'s are cloned or more messages are being sent. Later implementations (crossbeam, flume) do not necessarily do this.

> Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.

This is quite unfortunate. I wonder if somewhere either a `block_on` or `block_in_place` is already being used.

Reply:
> This is quite unfortunate.

Why do we need to use `tokio::sync::oneshot` if it provides not performance advantages comparing with `std::sync::mpsc`
where there are no such problems?